classes/manager.php

```diff
@@ -163,12 +168,190 @@ public static function get_reasons(float $duration): int {
+        ]);
+        return $count >= $quota;
+    }
+
+    /**
+     * Quota for this page, for this profile type (e.g. REASON_SLOW) has been reached.
+     *
+     * @param int reason
+     * @return bool whether or not the quota is filled.
+     */
+    public static function has_filled_page_and_reason_quota(string $request, int $reason): bool {
+        global $DB;
+        $reasonstr = self::REASON_STR_MAP[$reason];
+        $quota = (int) get_config('tool_excimer', "num_' . $reasonstr . '_by_page");
+
+        // Get and set cache.
+        $reasons = $DB->sql_bitand('reason', $reason);
+        $sql = "SELECT count(*)
+                  FROM {tool_excimer_profiles}
+                 WHERE $reasons != ?
+                       AND request = ?";
+        $count = $DB->count_records_sql($sql, [
+            self::REASON_NONE,
+            $request,
+        ]);
+        return $count >= $quota;
+    }
+
+    /**
+     * Checks the quotas, and returns the best value for the min duration a
+     * profile should be, before it should be saved.
+     *
+     * This will check quotass per page first, then check the more broad quotas, because it only
+     * matters if the page quota has been exceeded, e.g. after the broad quota
+     * has been reached.
+     *
+     * Assuming limits of page=5 and overall=10 (unlikely to be less than the
+     * page quota). The following behaviour is based on this assumption, that
+     * the page limit is less than the other limit (the reason limit or overall
+     * limit in this example).
+     *
+     * With all examples, the check should look at the profile duration based on
+     * the quota that's filled, and if both are filled then it should look at
+     * the pagequota values before adding a new profile.
+     *
+     * Examples:
+     * [quota filled, pagequota filled]:
+     * Should look and add new profiles based on pagequota, as this is the limiting factor before it can be saved.
+     *
+     * [quota filled, pagequota notfilled]:
+     * Should be based on (overall) quota filled.
+     *
+     * [quota notfilled, pagequota filled]:
+     * Should be based on the page quota.
+     *
+     * [quota notfilled, pagequota notfilled]:
+     * Should be based on configuration thresholds/limits.
+     *
+     * @return float the minimum duration required, for a profile to be stored with the REASON_AUTO reason.
+     */
+    public static function get_min_duration_for_reason_slow(): float {
+        // Quota for this page, for this profile type (e.g. REASON_SLOW) has been reached.
+        $request = profile::get_request();
+
+        // Get the cached timings for the fastest of the stored profiles, to
+        // ensure anything faster than this does not get stored iif the quota is
+        // reached. This cache should be reset when, a new profile is
+        // stored/deleted, or settings have changed.
+        if (self::has_filled_page_and_reason_quota($request, self::REASON_AUTO)) {
```

Comment:
Would this be simpler and use lets lookups if you did get_min_duration_for_reason() first (which should not require any extra read, see above) and only then in the small number of cases that exceed that then check the per requests cached limit.

I also don't think we need to check the quota in either case. If we only ever store the exact Nth quota we implicitly get this for free

Reply:
Yes based on the above feedback, this would reduce the overall cost of lookups assuming the 
- request quota is not reached
- and the profile is rejected because it is not slow enough to be counted as a profile to be stored.

Because now the cost of the system (per reason) quota is free/zero as mentioned above, if stored via get/set_config which is always loaded fully apparently. Inversely, If the costs of the 2 checks were the same, it would have made more sense to check the conditions required to whether or not it should be stored.

I wonder if we can cheat and also store the 'quota filled' request durations in get/set_config, and as mentioned above do the check if this key exists. The gain here is that we can avoid another single cache read e.g. for a slower than usual request endpoint (slower than the fastest stored profile for the reason, but faster than the fastest stored profile for the reason AND request), where the both quotas have been filled up and the profiler is deciding if the profile is worth adding, at the expense of having durations for N requests (or an arbitrary cap) stored in the plugin config as well.